File: app/src/main/java/com/example/earthwallet/chain/Staking.kt
package network.erth.wallet.chain

import com.google.protobuf.Any as ProtoAny
import cosmos.base.v1beta1.CoinOuterClass
import cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward
import cosmos.staking.v1beta1.MsgBeginRedelegate
import cosmos.staking.v1beta1.MsgCancelUnbondingDelegation
import cosmos.staking.v1beta1.MsgDelegate
import cosmos.staking.v1beta1.MsgUndelegate
import network.erth.wallet.Constants
import org.json.JSONObject

/** Native x/staking + x/distribution queries and messages. */
object Staking {

    data class Validator(
        val operator: String,
        val moniker: String,
        val tokens: String,
        /** Commission rate as a fraction, e.g. 0.10 for 10%. */
        val commission: Double = 0.0,
    )
    data class Delegation(val validator: String, val amount: String)
    /** A single native unbonding entry (funds auto-return to balance at [completionTime]). */
    data class UnbondingEntry(
        val validator: String,
        val balance: String,
        val completionTime: String,
        /**
         * Height the entry was created at. Cancelling addresses an entry by
         * (validator, creationHeight) — unbonding entries have no id — so this
         * must be carried through or the cancel cannot be built.
         */
        val creationHeight: Long,
    )

    /** Stake in flight between validators: still bonded, but locked until it matures. */
    data class RedelegationEntry(
        val src: String,
        val dst: String,
        val balance: String,
        val completionTime: String,
    )

    private fun uerth(amount: String) =
        CoinOuterClass.Coin.newBuilder().setDenom(Constants.UERTH_DENOM).setAmount(amount).build()

    // --- queries ---

    fun bondedValidators(): List<Validator> {
        val (code, body) = EarthRest.get(
            "/cosmos/staking/v1beta1/validators?status=BOND_STATUS_BONDED&pagination.limit=200"
        )
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("validators") ?: return emptyList()
        val out = ArrayList<Validator>(arr.length())
        for (i in 0 until arr.length()) {
            val v = arr.getJSONObject(i)
            out.add(
                Validator(
                    operator = v.getString("operator_address"),
                    moniker = v.optJSONObject("description")?.optString("moniker", "") ?: "",
                    tokens = v.optString("tokens", "0"),
                    commission = v.optJSONObject("commission")
                        ?.optJSONObject("commission_rates")
                        ?.optString("rate", "0")?.toDoubleOrNull() ?: 0.0,
                )
            )
        }
        return out
    }

    fun delegations(delegator: String): List<Delegation> {
        val (code, body) = EarthRest.get("/cosmos/staking/v1beta1/delegations/$delegator")
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("delegation_responses") ?: return emptyList()
        val out = ArrayList<Delegation>(arr.length())
        for (i in 0 until arr.length()) {
            val d = arr.getJSONObject(i)
            out.add(
                Delegation(
                    validator = d.getJSONObject("delegation").getString("validator_address"),
                    amount = d.getJSONObject("balance").getString("amount"),
                )
            )
        }
        return out
    }

    /** Total uerth bonded across the whole network (staking pool bonded_tokens). */
    fun totalBonded(): String {
        val (code, body) = EarthRest.get("/cosmos/staking/v1beta1/pool")
        if (code !in 200..299) return "0"
        return JSONObject(body).optJSONObject("pool")?.optString("bonded_tokens", "0") ?: "0"
    }

    /** In-progress unbonding entries for a delegator (native staking auto-releases at completion). */
    fun unbondingDelegations(delegator: String): List<UnbondingEntry> {
        val (code, body) = EarthRest.get("/cosmos/staking/v1beta1/delegators/$delegator/unbonding_delegations")
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("unbonding_responses") ?: return emptyList()
        val out = ArrayList<UnbondingEntry>()
        for (i in 0 until arr.length()) {
            val resp = arr.getJSONObject(i)
            val validator = resp.getString("validator_address")
            val entries = resp.optJSONArray("entries") ?: continue
            for (j in 0 until entries.length()) {
                val e = entries.getJSONObject(j)
                out.add(
                    UnbondingEntry(
                        validator = validator,
                        balance = e.optString("balance", "0"),
                        creationHeight = e.optString("creation_height", "0").toLongOrNull() ?: 0L,
                        completionTime = e.optString("completion_time", ""),
                    )
                )
            }
        }
        return out
    }

    /** Total pending uerth rewards across all validators. */
    fun totalRewards(delegator: String): String {
        val (code, body) = EarthRest.get("/cosmos/distribution/v1beta1/delegators/$delegator/rewards")
        if (code !in 200..299) return "0"
        val total = JSONObject(body).optJSONArray("total") ?: return "0"
        for (i in 0 until total.length()) {
            val c = total.getJSONObject(i)
            if (c.getString("denom") == Constants.UERTH_DENOM) {
                // rewards are DecCoins (may be fractional); truncate to integer uerth
                return c.getString("amount").substringBefore(".")
            }
        }
        return "0"
    }

    /** In-progress redelegations for a delegator. */
    fun redelegations(delegator: String): List<RedelegationEntry> {
        val (code, body) = EarthRest.get("/cosmos/staking/v1beta1/delegators/$delegator/redelegations")
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("redelegation_responses") ?: return emptyList()
        val out = ArrayList<RedelegationEntry>()
        for (i in 0 until arr.length()) {
            val resp = arr.getJSONObject(i)
            val red = resp.optJSONObject("redelegation")
            val entries = resp.optJSONArray("entries") ?: continue
            for (j in 0 until entries.length()) {
                val e = entries.getJSONObject(j)
                out.add(
                    RedelegationEntry(
                        src = red?.optString("validator_src_address", "") ?: "",
                        dst = red?.optString("validator_dst_address", "") ?: "",
                        balance = e.optString("balance", "0"),
                        completionTime = e.optJSONObject("redelegation_entry")
                            ?.optString("completion_time", "") ?: "",
                    )
                )
            }
        }
        return out
    }

    // --- messages ---

    fun msgDelegate(delegator: String, validator: String, amountUerth: String): ProtoAny =
        EarthTx.anyOf(
            "/cosmos.staking.v1beta1.MsgDelegate",
            MsgDelegate.newBuilder()
                .setDelegatorAddress(delegator).setValidatorAddress(validator).setAmount(uerth(amountUerth))
                .build()
        )

    fun msgUndelegate(delegator: String, validator: String, amountUerth: String): ProtoAny =
        EarthTx.anyOf(
            "/cosmos.staking.v1beta1.MsgUndelegate",
            MsgUndelegate.newBuilder()
                .setDelegatorAddress(delegator).setValidatorAddress(validator).setAmount(uerth(amountUerth))
                .build()
        )

    /**
     * Move stake between validators without unbonding — it keeps earning, with no
     * 21-day gap. The chain refuses to redelegate stake that is already in flight,
     * and caps concurrent entries between any validator pair.
     */
    fun msgBeginRedelegate(delegator: String, src: String, dst: String, amountUerth: String): ProtoAny =
        EarthTx.anyOf(
            "/cosmos.staking.v1beta1.MsgBeginRedelegate",
            MsgBeginRedelegate.newBuilder()
                .setDelegatorAddress(delegator)
                .setValidatorSrcAddress(src)
                .setValidatorDstAddress(dst)
                .setAmount(uerth(amountUerth))
                .build()
        )

    /**
     * Cancel an in-progress unbonding, returning the stake to the same validator.
     * Partial cancels are allowed; the remainder keeps its original schedule.
     */
    fun msgCancelUnbonding(
        delegator: String,
        validator: String,
        amountUerth: String,
        creationHeight: Long,
    ): ProtoAny =
        EarthTx.anyOf(
            "/cosmos.staking.v1beta1.MsgCancelUnbondingDelegation",
            MsgCancelUnbondingDelegation.newBuilder()
                .setDelegatorAddress(delegator)
                .setValidatorAddress(validator)
                .setAmount(uerth(amountUerth))
                .setCreationHeight(creationHeight)
                .build()
        )

    fun msgWithdrawReward(delegator: String, validator: String): ProtoAny =
        EarthTx.anyOf(
            "/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward",
            MsgWithdrawDelegatorReward.newBuilder()
                .setDelegatorAddress(delegator).setValidatorAddress(validator)
                .build()
        )
}

File: ios/EarthCore/Sources/EarthCore/Chain/Staking.swift
import Foundation

/// Native x/staking and x/distribution.
public enum Staking {

    public struct Validator: Sendable, Equatable, Identifiable {
        /// The operator address. A validator can change its moniker, so the
        /// list would reorder under a name-keyed identity.
        public var id: String { operatorAddress }

        public let operatorAddress: String
        public let moniker: String
        public let tokens: String
        /// Commission as a fraction — 0.10 is 10%.
        public let commission: Double
    }

    public struct Delegation: Sendable, Equatable, Identifiable {
        public var id: String { validator }

        public let validator: String
        public let amount: String
    }

    /// One unbonding entry. Funds return to the balance on their own at
    /// `completionTime`.
    public struct UnbondingEntry: Sendable, Equatable {
        public let validator: String
        public let balance: String
        public let completionTime: String
        /// Cancelling addresses an entry by (validator, creationHeight) —
        /// unbonding entries have no id — so this has to be carried through or
        /// the cancel cannot be built.
        public let creationHeight: Int64
    }

    /// Stake in flight between validators: still bonded, but locked until it
    /// matures.
    public struct RedelegationEntry: Sendable, Equatable {
        public let source: String
        public let destination: String
        public let balance: String
        public let completionTime: String
    }
}

public extension EarthClient {

    func bondedValidators() async -> [Staking.Validator] {
        guard let json = try? await rest.get(
            "/cosmos/staking/v1beta1/validators?status=BOND_STATUS_BONDED&pagination.limit=200"
        ) else { return [] }
        return json.validators.array.compactMap { v in
            guard let op = v.operator_address.string else { return nil }
            return Staking.Validator(
                operatorAddress: op,
                moniker: v.description.moniker.string(default: ""),
                tokens: v.tokens.string(default: "0"),
                commission: v.commission.commission_rates.rate.double(default: 0)
            )
        }
    }

    func delegations(_ delegator: String) async -> [Staking.Delegation] {
        guard let json = try? await rest.get("/cosmos/staking/v1beta1/delegations/\(delegator)")
        else { return [] }
        return json.delegation_responses.array.compactMap { d in
            guard let validator = d.delegation.validator_address.string else { return nil }
            return Staking.Delegation(validator: validator, amount: d.balance.amount.string(default: "0"))
        }
    }

    /// Total uerth bonded across the network.
    func totalBonded() async -> String {
        guard let json = try? await rest.get("/cosmos/staking/v1beta1/pool") else { return "0" }
        return json.pool.bonded_tokens.string(default: "0")
    }

    func unbondingDelegations(_ delegator: String) async -> [Staking.UnbondingEntry] {
        guard let json = try? await rest.get(
            "/cosmos/staking/v1beta1/delegators/\(delegator)/unbonding_delegations"
        ) else { return [] }
        return json.unbonding_responses.array.flatMap { response -> [Staking.UnbondingEntry] in
            guard let validator = response.validator_address.string else { return [] }
            return response.entries.array.map { e in
                Staking.UnbondingEntry(
                    validator: validator,
                    balance: e.balance.string(default: "0"),
                    completionTime: e.completion_time.string(default: ""),
                    creationHeight: e.creation_height.int64(default: 0)
                )
            }
        }
    }

    func redelegations(_ delegator: String) async -> [Staking.RedelegationEntry] {
        guard let json = try? await rest.get(
            "/cosmos/staking/v1beta1/delegators/\(delegator)/redelegations"
        ) else { return [] }
        return json.redelegation_responses.array.flatMap { response -> [Staking.RedelegationEntry] in
            let redelegation = response.redelegation
            return response.entries.array.map { e in
                Staking.RedelegationEntry(
                    source: redelegation.validator_src_address.string(default: ""),
                    destination: redelegation.validator_dst_address.string(default: ""),
                    balance: e.balance.string(default: "0"),
                    completionTime: e.redelegation_entry.completion_time.string(default: "")
                )
            }
        }
    }

    /// Total pending uerth rewards across all validators.
    ///
    /// Rewards are DecCoins and may be fractional, so the fraction is dropped
    /// rather than rounded — the chain pays out the truncated amount.
    func totalRewards(_ delegator: String) async -> String {
        guard let json = try? await rest.get(
            "/cosmos/distribution/v1beta1/delegators/\(delegator)/rewards"
        ) else { return "0" }
        for coin in json.total.array where coin.denom.string == Constants.gasDenom {
            let amount = coin.amount.string(default: "0")
            return String(amount.prefix(while: { $0 != "." }))
        }
        return "0"
    }

    // --- messages ---

    private func uerth(_ amount: String) -> Coin {
        Coin(denom: Constants.gasDenom, amount: amount)
    }

    func msgDelegate(delegator: String, validator: String, amountUerth: String) -> ProtoAny {
        Msg.Delegate(delegator: delegator, validator: validator, amount: uerth(amountUerth))
            .asAny(typeURL: Msg.Delegate.typeURL)
    }

    func msgUndelegate(delegator: String, validator: String, amountUerth: String) -> ProtoAny {
        Msg.Undelegate(delegator: delegator, validator: validator, amount: uerth(amountUerth))
            .asAny(typeURL: Msg.Undelegate.typeURL)
    }

    /// Move stake between validators without unbonding — it keeps earning, with
    /// no 21-day gap. The chain refuses to redelegate stake already in flight,
    /// and caps concurrent entries between any validator pair.
    func msgBeginRedelegate(
        delegator: String,
        source: String,
        destination: String,
        amountUerth: String
    ) -> ProtoAny {
        Msg.BeginRedelegate(
            delegator: delegator,
            source: source,
            destination: destination,
            amount: uerth(amountUerth)
        ).asAny(typeURL: Msg.BeginRedelegate.typeURL)
    }

    /// Cancel an in-progress unbonding, returning the stake to the same
    /// validator. Partial cancels are allowed; the remainder keeps its schedule.
    func msgCancelUnbonding(
        delegator: String,
        validator: String,
        amountUerth: String,
        creationHeight: Int64
    ) -> ProtoAny {
        Msg.CancelUnbondingDelegation(
            delegator: delegator,
            validator: validator,
            amount: uerth(amountUerth),
            creationHeight: creationHeight
        ).asAny(typeURL: Msg.CancelUnbondingDelegation.typeURL)
    }

    func msgWithdrawReward(delegator: String, validator: String) -> ProtoAny {
        Msg.WithdrawDelegatorReward(delegator: delegator, validator: validator)
            .asAny(typeURL: Msg.WithdrawDelegatorReward.typeURL)
    }
}
